### ISSUE TITLE
P4_16: Allow 'bit-vector structs' to be used everywhere that bit<W> type can

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4-spec#383]
Or at least almost everywhere.

By a 'bit-vector struct' is meant the same thing as described in this issue: https://github.com/p4lang/p4-spec/issues/342

Motivation: structs and nested structs can make it noticeably more convenient to write code that wants to bundle together bit<W> fields that occur together into a unit.

The proposal is to allow bit-vector structs to be used in just about any place that a single bit<W> type value is permitted today.  Some of these may already be permitted by the existing P4_16 spec and open source tools, but an explicit mention in the spec might help ensure they gain wide adoption in P4-based systems.

Proposed that bit-vector structs should be supported:
+ as parameters of actions
+ as keys in tables
+ as parameters of extern methods (e.g. data parameter of Register.write)
+ as return values of extern methods (e.g. return value of Register.read)
+ allow them to be cast to the type `bit<W>` where W is the total width of all fields in the bit-vector struct (including nested structs)
+ allow `bit<W>` type to be cast to type of a bit-vector struct with the same total width in bits

**Comment by mbudiu-vmw** [comment]
This is not trivial. It has important consequences for the control-plane api. Also, casting may be very expensive on some architectures (e.g. Software where p4 structs are mapped to c structs which have different allignment rules).

**Comment by jnfoster** [comment]
I don't think there are necessarily complex run-time costs associated with this feature. The compiler can recognize these type equalities but use a canonical representation (e.g., flat bits) internally. (The one source of additional complexity I see is what you point out: ideally P4Runtime would allow any of the bit types representations defined in the program to be used...)

**Comment by orenbenkiki** [comment]
@mbudiu-vmw, can you elaborate on the C `struct` issue?

Some points that come to mind:

1.  Andy's proposal applies only to structures containing only `int<W>` and `bit<W>` (possibly nested) fields, which I expect to be converted to C bit fields in (possibly nested) `struct` with a `__packed__` attribute (or the equivalent in your favorite compiler). Is it an issue to expect such `struct` types to be packed? 

2.  The size of a C `struct` must be an integer number of bytes. That is, in C it would be impossible to _naively_ pack two bit-vector-structs of size 12 bits inside a container bit-vector-struct whose size is 24 bits. It would be, however, possible to represent this kind of layout in C with additional compiler work: Since P4 has no notion of a "pointer", the compiler can emit the container `struct` as a flat list of fields and perform the necessary code transformation to access the appropriate fields. Is it an issue to expect the compiler to do this work for us?

3. The efficiency of code dealing with packed `struct` types would be lower than code dealing with `struct` types that are aligned with additional padding. Is this efficiency impact important enough to warrant explicitly annotating, in the P4 source code, which `struct` types are "bit-vector" and which aren't?

4. If efficiency is an issue, we could annotate such `struct` types with a `@packed` attribute, or  use a new `bits_struct` keyword, or use some other explicit expression of the intent. Would making the bit-vector-struct concept explicit lay some of the concerns to rest?

**Comment by orenbenkiki** [comment]
@jnfoster, allowing the internal run-time representation to include hidden padding would make it much easier to convert a bit-vector-struct to a C `struct`.

That said, assuming that one would be able to convert a bit-vector-struct into/from a bit-vector, then that compiler will need to generate appropriate conversion code to deal with any "hidden" padding. This seems like a valid implementation strategy - as long as no code makes any assumptions about the internal representation. The worry here is that, for example, external C code which interacts with the P4 code, will directly access the C `struct` representation and therefore become dependent on the specific padding. Is this something we can prevent by warning people from such dependencies, possibly by referring to C99 6.7.2.1#10 and similar standards?

**Comment by ChrisDodd** [comment]
I think Mihai is complaining primarily about the cast between struct and bitvector operations.  The reason is that on some platforms it may be expensive to do operations on values that are arbitrary bitfields not byte or word aligned.  On such platforms structs of bits will generally be broken up with one field per byte/word rather than packed together, so casting between them and bit<N> values will involve packing and unpacking.

They already need to do this for headers, as the wire format for headers must be packed, but most headers tend to be byte aligned, and there is probably extra hardware in the parser/deparser to deal with this that may not be present in whatever hardware is managing match-action controls; this division of labor is a big reason why P4 has parser and deparser separate from the main match-action control.

**Comment by orenbenkiki** [comment]
Yes, conversion to/from a bit vector might be more expensive for some structures for some hardware. This might be a good reason to explicitly annotate bit-vector-structures,

That said, currently match/action code is free to use header types, including headers that are not a result of parsing, so the compiler already has to deal today with such issues even in such match/action code - the separation between parser/deparser and match/action doesn't seem to be that sharp in this respect.

Perhaps the problem is due to the concepts of "header" vs. "structure" - headers can be viewed as bit-vectors, and structures can't. If that were the only difference, there wouldn't be a problem, as the code could just use a header instead of a bit-vector-structure.

However, headers carry an additional "validity" bit with them, while structures do not. In principle, a sufficiently smart compiler might be able to detect when this bit is irrelevant and optimize it away, at least in common cases. Probably.

The killer difference is that headers can't be nested (why?). This has a huge negative impact on their usefulness. If headers could be freely nested, there would be much less need for bit-vector-structures.

**Comment by jafingerhut** [comment]
@orenbenkiki Allowing headers to nest would raise its own set of questions that would need answering in the context of the P4_16 language.

I can easily imagine that if one did an extract() on a header that contained nested headers, then all of the headers involved would become valid.

Doing hdr.setInvalid() on the "top level" header then makes all sub-header invalid, too?  Seems to be a reasonable choice.

What about doing hdr.setInvalid() on a 'sub-header', without doing so on the top level header?  What if you tried to then emit() the top level header in that state?  What should be emitted?

For reasons like that, the 'bit-vector struct' idea seems less problematic, at least from the limited thinking I have done on it so far.

**Comment by orenbenkiki** [comment]
Yes, nesting headers would complicate their definition. I agree that allowing for a variant of structs is cleaner, given the current state of affairs (it is way too late for P4 to use something like "Optional<...>", more's the pity).

I think that explicitly annotating bit-vector-struct addresses the concerns about the performance issues. It is the same concept as `__packed__` C/C++ structs - including the performance issues. I think that carrying the "packed" terminology to P4 makes sense.

**Comment by mbudiu-vmw** [comment]
P4 attempts to be as much as possible target-independent. It does not discuss anything about the memory representation of various data structures; for headers it does mention the wire representation produced by emit and extract. This is why there is no need to talk about packing.

Each back-end can make choices about the actual implementation that suitable for the target. The only C back-end I wrote is the EBPF back-end, which is probably not very efficient, but it is a plausible implementation.  There each struct field is allocated an integer number of bytes. In this implementation converting a "struct of bit-vectors" to a bit-vector would be a very expensive operation, requiring lots of shifts and masks.

**Comment by hesingh** [comment]
Please also see https://github.com/p4lang/p4c/issues/562 

where this work is being continues

**Comment by jnfoster** [comment]
Discussed on 11/5/2018 LDWG. Seems like a promising feature enhancement, but no owner for now.

**Comment by hesingh** [comment]
Please note that another PR in #562 is tracking support for nested bit-vector structs.  #562 is not only going to add the support for bmv2, but any other backend, if so asked for.  #562 was on hold until Mihai completes structure initialization which helps the flattening midend pass.  Any nested struct support is tied to also supporting a midend pass to flatten nested structs.

**Comment by hesingh** [comment]
A specification is provided for how to add nested bit-vector structs support to P4-16 language and compiler.

A Use case is provided below:

A programmable switching asic we developed a p4c backed for, uses union of data structures.  P4 only supports union of P4 headers via a header_union. Thus, our backend data structures are P4 headers. Further, the asic hardware sub-element (Mux Comparator) uses a data structure with bit-fields and structures.  This use case demands structures in headers.  A Mux Comparator, common to many switching asics, performs comparisons in parallel.  The hardware has finite number of instructions to compare data. If a data structure is un-rolled, the Mux Comparator runs out of instructions.  The Mux has enough width but not enough instructions to compare.  This is why nested bit-vector structs work with this Mux.  We used the simplest solution by changing p4c frontend to support nested bit-vector for struct and P4 header along with Cast support.  The backend didn't have to change. 
The nested structures or headers are expected as parameters to at least P4 Control and table key. Alternatives may exist to solve the problem, but may complicate the backend.  A structure overlay creates more complexity in the backend vs. a nested bit-vector structure. Also, nested bit-vector struct requires Cast support as well. 

Specifically, p4c should be modified to support:
1.	Nested bit-vector structures. 
2.	Nested bit-vector structures in P4 header.
3.	Casting of nested bit-vector structure to bit-field of same width or vice versa.
4.	Item 1 and 2 as parameters in P4 Action, Control, and Parser.
5.	The bmv2 backend changes to support items 1 and 2.
6.	A P4 header_union with headers including nested bit-vector structure(s).
7.	Only P4-16.

A nested header is not supported.  

Additionally, a table key or any other P4 type used by P4Runtime cannot use items 1 and 2 above.  This rule allows control-plane API generation to not be impacted.  P4Runtime supports Tables, PSA Action Profiles, Meters, Counters, Packet Replication, Parser Values, Registers, Digests, and Externs.

**Comment by jnfoster** [comment]
Hi Hemant,

Speaking and an individual and not LDWG co-chair: I generally support
generalizing the type system to allow constructs and features to be
combined in more ways provided it doesn't complicate implementations on
broad classes of targets.

Having said that, there are a lot of changes here, and we do need to
prioritize. It sounds to me like your use case is actually driven by the
fact that it eases building a backend for a specific target. If that's
accurate, will you be driving the development of a prototype implementation
on p4c and bmv2, example programs, and a PR on the specification document?
I think if you come to the group with those three things, people would be
glad to take these changes.

I do have one technical concern: items 2 and 6 propose putting structs as
members of P4 headers and header unions. This is a topic we've discussed
before: it makes the extract/emit operations more difficult to specify
and/or forces us to commit to a particular serialization of structs, which
is something various people find less attractive.

-N

**Comment by jnfoster** [comment]
I like @akeep's proposal: we allow many types to be treated as
serializable, possibly with extra annotations. Then the requirement for
headers would simply be that it (and therefore all of its members) must be
serializable. That seems like a more elegant way to make the distinction
than in vs. not-in a header.

**Comment by hesingh** [comment]
@jnfoster: In regards to "development of a prototype implementation on p4c and bmv2, example programs, and a PR on the specification document? I think if you come to the group with those three things, people would be glad to take these changes.",  

It makes sense and I can do that, thanks.

In regards to extract and emit getting complicated if a P4 header includes a nested bit-vector struct, let's see.  I understand extract and emit work on a P4 header.  A P4 struct is not any different that a P4 header which includes a hidden isValid() bit.  So, why is it complicated to extend extract and emit to also support a P4 struct?  Sorry, if I missed anything.

I am willing to hold off on 6.  Please also note the text at the end of my note.  This text says to not use items 1 and 2 for constructs used by P4Runtime, so that api generation by p4c is not impacted.

**Comment by hesingh** [comment]
@jnforster:  Is there a link where I can see @akeep's proposal?  thanks.

**Comment by jnfoster** [comment]
The main record I know of is in the past LDWG meeting minutes, which are on the Wiki.

**Comment by jafingerhut** [comment]
@hesingh "So, why is it complicated to extend extract and emit to also support a P4 struct?"

It isn't complicated in the simplest cases you are probably imagining, which is when a struct contains only fields of type `bit<W>`, or other structs that are themselves like that, nested arbitrarily.

Officially, P4_16 defines a bit ordering for extract and emit on headers.  It doesn't do so for structs.  It could, but doesn't.

If a struct contains only `bit<W>` or other structs that are similar, it is straightforward to imagine extending the definition of emit/extract to headers containing such structs.

It is not so straightforward if the struct contains fields of other types, e.g. enum (without a backing `bit<W>` type), error, headers, header_union, header stacks, etc.  If:

+ a struct that only contained other "easily serializable" things could become "easily serializable" because it only contained such things, and
+ structs containing other things like type `error` fields that are not serializable (or at least no satisfactory way has been defined in the P4_16 language spec yet)

then structs of the first kind I think would be nice if they could be included in headers.

**Comment by hesingh** [comment]
@jafingerhut :  Agree.  Note, the specification I entered today to this PR says "nested bit-vector struct".  I didn't envision supporting any other type such as error, header_union, etc. in a nested struct.  If only a nested bit-vector struct" is included in a header, then the data struct is serializable.  It's not complicated to add bit ordering support to a bit-vector struct for extract and emit. 

@jnfoster: I have read akeep's note: https://github.com/akeep/p4-proposals/blob/master/serializable.pdf

I don't see any problem with adding nested bit-vector struct to a header and not have this data struct serializable.

**Comment by orenbenkiki** [comment]
> @hesingh "So, why is it complicated to extend extract and emit to also support a P4 struct?"

> Officially, P4_16 defines a bit ordering for extract and emit on headers. It doesn't do so for structs. It could, but doesn't.

Actually it does? Both https://p4.org/p4-spec/docs/P4-16-v1.0.0-spec.html#sec-deparse-insert and https://p4.org/p4-spec/docs/P4-16-v1.1.0-draft.html#sec-deparse-insert say that `emit` **can** be applied to a struct, in which case it is applied to each field in order, and this is only valid if no (nested) field is an `enum` or an `error`.

> If a struct contains only ...[types that have bit representation]...I think would be nice if they could be included in headers.

Yes, there is absolutely no reason why it isn't valid to place a struct-that-can-be-emitted inside a header. The spec tries to pretend structs don't have a "bit representation" but at the same time does allow them to be emitted as bits. I always found it to be inconsistent. I feel that a better choice would have been making the concepts of "stuff that has bit representation" and "stuff that has an is-valid bit" to be orthogonal, instead of conflating both concepts into a single "header" type. That is, having something like `std::optional<T>` in C++.

**Comment by hesingh** [comment]
@orenbenkiki: thanks for the emit reference that shows emit supports a struct.  However, we still need to change extract to support a struct - extract only supports a header.

**Comment by orenbenkiki** [comment]
@hesingh: You are right. What we need is for the spec to fully accept that (restricted) structs have a bit representation and embrace the implications; in particular, allow the full set of operations to them - place them in headers and unions, extract them, etc. as opposed to merely allowing to emit them. The current set of restrictions seems to be arbitrary and is probably due to historical reasons rather than technical reasons.

**Comment by hesingh** [comment]
Note, in the specification I added to this PR today, I have said
"Additionally, a table key or any other P4 type used by P4Runtime cannot use items 1 and 2 above. This rule allows control-plane API generation to not be impacted. P4Runtime supports Tables, PSA Action Profiles, Meters, Counters, Packet Replication, Parser Values, Registers, Digests, and Externs."

On more thought, I see jafingerhut has asked for nested bit-vector support for table keys etc. when he first authored this PR.  If a header/struct includes a nested bit-vector struct and this header/struct can be serialized, the control-plane in P4Runtime can also serialize the same header/struct.  Thus, it makes sense to relax the P4Runtime requirement.  Of course, then the P4Runtime controller and switch agent would also need to support nested bit-vector struct and header.   Further, if a p4c backend uses nested-bit vector struct and header, the backend must not enable the new midend pass to flatten nested bit-vector struct - otherwise the control-plane API is out of sync with target firmware.

**Comment by jafingerhut** [comment]
P4Runtime API v1.0 will restrict the supported types of action parameters and table keys to be `bit<W>`, because some controllers and agent software in active development has that restriction.

P4Runtime API v1.0 does explicitly allow and define a Protobuf serialization format for all P4_16 types, including nested structs, and these should be supported for Digest extern messages, and hopefully also Register extern array element contents.  I believe it is expected in the future that this same format could also be used for table keys and action parameters, even if P4_16 never defines a way to serialize such types.

**Comment by hesingh** [comment]
This is the public code workspace to see the code changes.  Few legacy cases need to be fixed - once these test cases are fixed, I will pull a request.

https://github.com/hesingh/p4c-1

What is completed so far:

1. extract() functionality has be extended to support the special struct.
2. p4c has been changed to support the new struct including Cast functionality.
3. Added test case to test emit() with struct and testing has passed.  

Please see this URL for code changes: 
https://github.com/hesingh/p4c-1/commit/02acc32a44a2d27910842877aed1474f29030e04

**Comment by hanw** [comment]
@hesingh, there are a number of changes to the spec in your commit. It might be easier to review if we could break them apart into smaller changes, each for one change in the spec. For example, I could argue that the `clone_header_t` has a well-defined bit layout, given the fact that the struct is used inside a header and the header dictates the bit layout.
```
struct clone_field_list {
   bit<16> f0;
   bit<8> f1;
}
header clone_header_t {
   bit<8> tag;
   clone_field_list fl;
}
```

On the other hand, the following example does not have a well-defined bit layout, because the "outer" struct cannot enforce the order between `tag` and `fl`.
```
header clone_field_list {
   bit<16> f0;
   bit<8> f1;
}
struct clone_header_t {
   bit<8> tag;
   clone_field_list fl;
}
```

From my experience, if we can support the first use case, even if we do not change the semantics of `packet_out.emit()`, we can already have a lot more flexibility in P4.

Can you give us an example that shows why `packet_in.extract()` has to be changed to support `struct`, if we allow `struct` in a `header`?

**Comment by hesingh** [comment]
@hanw: If you see the specification I added to this Issue 13 days ago, the first two items are:

1. Nested bit-vector structures. 
2. Nested bit-vector structures in P4 header.

The case you raised for not being serializable, is a case of nested header inside struct - this case is not supported by the specification of this Issue.  I can add a new test case to show this case will fail compiling with my changes. 

In regards to an example why packet.extract() has to be changed to support struct, I just committed a modified test case - see p4_16_samples/extract_struct.p4.  Further, it makes sense to maintain symmetry.  The P4-16 spec and the compiler support both header and struct with emit().  For completeness, I think, even extract makes sense to be modified to also support a struct.

**Comment by hesingh** [comment]
The first three items are a must-have.  The third item is included below.

3. Casting of nested bit-vector structure to bit-field of same width or vice versa.

It's no use developing 1 and 2 if 3 is not supported.  See my comment in the following Issue regarding Cast.

https://github.com/p4lang/p4-spec/issues/342

I can hold off on extract() so that folks can mull over it and provide comments for why a bit-vector struct should not be supported by extract().

**Comment by hanw** [comment]
IIRC, the reason that `emit()` is allowed to support struct is for convenience, such that user can write the deparser code more succinctly, rather than spelling out all the headers inside a struct. For example:
```
header data_a_h {
  bit<32> da;
  bit<32> db;
}
header data_b_h {
  bit<16> da;
  bit<16> db;
}
struct headers_t {
  data_a_h data_a;
  data_b_h data_b;
}
control SwitchIngressDeparser(
    packet_out pkt,
    inout headers_t hdr,
    in user_metadata_t meta) {
    apply {
        pkt.emit(hdr);
    }
}
``` 

For `extract()`, there is probably less need for such syntax sugar.  Further, suppose we allow `struct` inside a `header`, you can always extract the special struct by putting it inside a header.

Changing `extract()` to support struct also means that we would need to change the following text in P4-16 spec. It seems like a pretty significant change to the spec to me, in terms of how the programmer would think about the difference between header and metadata when they program in P4.
```
P4's parser language provides an extract method that can be used to “fill in” the fields of a header from a network packet, as described in Section 12.8. The successful execution of an extract operation also sets the validity bit of the extracted header to true. 
```

**Comment by hanw** [comment]
@hesingh 
> The first three items are a must-have. The third item is included below.
>    Casting of nested bit-vector structure to bit-field of same width or vice versa.
> It's no use developing 1 and 2 if 3 is not supported. See my comment in the following Issue regarding Cast.

I do have another use case for 1 and 2, which does not use the cast feature. So I am in support of 1 and 2.

**Comment by hesingh** [comment]
Ah, good point about extract and setting the validity bit.  If only a struct is extracted, we'd be confused if the validity bit is set of not.  I will back out the extract code changes.  

In regards to case 3, I have already explained critical need for Cast soon as nested bit-vector structs are used.  Even Andy Fingerhut supports the Cast work.

**Comment by hanw** [comment]
I have no objection to case 3.

**Comment by jafingerhut** [comment]
@hesingh Be careful speaking for others.   I don't recall making any statements before about a desire or need for casting between structs and flat bit vectors.  Feel free to remind me where I made any comments on that topic, if I did, and I can go back and read it and see if "now me" agrees with "previous me". :-)

That said, I can't necessarily see any reason to object to explicit casts between nested 'bit-vector structs' and values of type `bit<W>` (assuming they have the same total size in bits).

It doesn't make sense to me why you say "there is no reason to do 1 and 2 without 3".  What goes wrong if you do 1 and 2 but not 3?

**Comment by hesingh** [comment]
@jafingerhut: Sorry, Andy.  However, at the top of this Issue your comments do mention Cast.  This is from your text.
"allow them to be cast to the type bit<W> where W is the total width of all fields in the bit-vector struct (including nested structs)"

I have already pointed to Issue 342 where I have given a clear case for Cast needed in many software situations.

**Comment by hanw** [comment]
Shall we make a PR for nesting bit vector struct inside header based on @hesingh's work, and consider adding that as an experimental feature? Given that there are more than one use cases for this feature. Casting from nested bit vector struct to bit type can be a separate PR?

My use case for nesting bit vector struct inside a header is as follows:

```
@flexible   /// I would like to be able to put an annotation on this struct.
struct switch_metadata_t {
    bit<9> port;
    bit<1> check;
}

header serialized_struct_t {
    switch_metadata_t meta;
}

struct switch_header_t {
    serialized_struct_t bridged_md;
    ethernet_h ethernet;
}

control SwitchIngress(
        inout switch_header_t hdr,
        inout switch_metadata_t ig_md) {

    action add_bridged_md() {
        hdr.bridged_md.setValid();
        hdr.bridged_md.meta = {ig_md.port, ig_md.check};
    }
}

control SwitchIngressDeparser(packet_out pkt,
                              inout switch_header_t hdr) {
    apply {
        pkt.emit(hdr.bridged_md);
        pkt.emit(hdr.ethernet);
    }
}

parser SwitchEgressParser(
        packet_in pkt,
        out switch_header_t hdr,
        out switch_metadata_t eg_md) {
    state start {
        pkt.extract(hdr.bridged_md);
        eg_md = hdr.bridged_md.meta;
        transition accept;
    }
}
```

**Comment by hesingh** [comment]
https://github.com/p4lang/p4c/pull/1623

**Comment by hanw** [comment]
Thanks! Mihai is most familiar with the type checker implementation, I have tagged him as a reviewer.

**Comment by hesingh** [comment]
A review comment from the PR is "test the frontend changes with a backend".  Of course, I have tested the frontend changes with a switching asic backend but the backend code is not in public-domain.  Thus, I have to test with the bmv2 backend.  

The bmv2 backend fails for the special struct inside a P4 header (item 2 in my specification) - bmv2 does not support such nesting.  Recently, Mihai has checked in a midend pass that flattens any nested structs (item 1) in a midend pass.  I have contacted Mihai via email to ask if we should modify the midend pass to flatten the special struct inside a P4 header.   If such a modification is done to the midend pass, then the bmv2 backend will not receive the nested struct inside the header.  Modifying the bmv2 backend to support the special struct in header is a bigger churn.

**Comment by mbudiu-vmw** [comment]
There are several passes that deal with nested structs in the BMv2 mid-end.
The bmv2 mid-end has NestedStructs, which eliminates nested structs that are local variables, CopyStructures, which eliminates assignments of values that are structs, and FlattenInterfaceStructs, which eliminates nested structs that appear within the interfaces of controls or parsers. All of these passes may need to be adapted. None of them expects a header to contain structs. There may be other places in the code where the fact that headers cannot contain structures is implicit. Hopefully we can locate them by writing more tests.

**Comment by hesingh** [comment]
Ok, Mihai, thanks.  

Other news is that I have already seen the CopyStructures will need mods to support Casting.  I will start at the NestedStructs midend pass to support a header with nested special struct.

**Comment by hesingh** [comment]
I adapted the flattenInterfaceStructs midend pass to support a special struct of item 2 in my specification.   With the adaptation, now the item 2 struct is getting flattened and bmv2 test cases passed.  Both extract and emit have also been added to the test case and their testing also passed.

**Comment by hesingh** [comment]
My machine did not have the behavioral-model repo.  I downloaded the repo since the PR for this Issue has to support bmv2 backend.  With the behavioral-model and using p4c code changes of the PR, lot of bm2 and other tests fail - about 203 in number.  The gross error relates to lines of code that use nested structs inside a P4 header.  The adapted flattened pass has flattened data structures, but the P4 program still uses the old, un-flattened data structures.  Thus a new midend pass has to find all uses of flattened data struct and change the used code.  Since multiple lines of code will be generated for one line of old code, bit-slices will be used.  See https://github.com/p4lang/p4c/issues/562 where I have explained this problem on May 25, 2018.  

`./p4c-bm2-ss -o bmv2/testdata/p4_16_samples/arith-bmv2.json /home/hemant/p4c-1/testdata/p4_16_samples/arith-bmv2.p4
Exit code  1
Error compiling
/home/hemant/p4c-1/testdata/p4_16_samples/arith-bmv2.p4(30): error: Structure header hdr does not have a field c
    { h.h.c = (bit<64>)(h.h.a + h.h.b); sm.egress_spec = 0; }
          ^
/home/hemant/p4c-1/testdata/p4_16_samples/arith-bmv2.p4(30): error: Structure header hdr does not have a field a
    { h.h.c = (bit<64>)(h.h.a + h.h.b); sm.egress_spec = 0; }
                            ^
/home/hemant/p4c-1/testdata/p4_16_samples/arith-bmv2.p4(30): error: Structure header hdr does not have a field b
    { h.h.c = (bit<64>)(h.h.a + h.h.b); sm.egress_spec = 0; }
                                    ^
`

**Comment by mbudiu-vmw** [comment]
This is the spec repository, we should keep discussions about the implementation in the implementation repository.

**Comment by hesingh** [comment]
> This is the spec repository, we should keep discussions about the implementation in the implementation repository.

Sorry, forgot about that.  Let me delete the note and add to the an implementation - thanks.

**Comment by ChrisDodd** [comment]
Re: emit/extract extensions:

The spec contains the line: `Architectures may impose additional constraints on header types---e.g., restricting headers to sizes that are an integer number of bytes.` in the header-types section.  The entire reason for this is support emit and extract -- generally the lower level will have have restrictions on the input/output of the data plane, such as it being a byte stream rather than a bit stream.  So if we want to support general emit/extract of "serializable" types, we'd probably need similar wording.

**Comment by hesingh** [comment]
@jafingerhut Now that the p4-spec has been changed to allow nested bit-vector struct, can we close this issue?

**Comment by jafingerhut** [comment]
Do you mean this PR? https://github.com/p4lang/p4-spec/pull/719

It hasn't actually been merged in yet, but I guess it is hopefully any day now that it will.  I tend to prefer to close issues after the changes have been merged in, rather than before.

**Comment by hesingh** [comment]
Yes, it's PR # 719.
